jax_distribution/systems/ff_ippo_rware.py

```diff
@@ -389,55 +512,169 @@ def learner_setup(env: Environment, config: Dict) -> Tuple[callable, RunnerState
+    """Initialise evaluator_fn, network, optimiser, environment and states."""
+    # Get available TPU cores.
+    n_devices = len(jax.devices())
+    # Get number of actions.
+    num_actions = int(eval_env.action_spec().num_values[0])
+
+    # Define network and vmap it over number of agents.
+    eval_network = ActorCritic(num_actions)
+    vmapped_eval_network_apply_fn = jax.vmap(
+        eval_network.apply,
+        in_axes=(None, 0),
+    )
+
+    # Pmap evaluator over cores.
+    evaluator, absolute_metric_evaluator = get_evaluator_fn(
+        eval_env, vmapped_eval_network_apply_fn, config
+    )
+    evaluator = jax.pmap(evaluator, axis_name="device")
+    absolute_metric_evaluator = jax.pmap(absolute_metric_evaluator, axis_name="device")
+
+    # Broadcast trained params to cores and split rngs for each core.
+    trained_params = jax.tree_util.tree_map(lambda x: x[:, 0, ...], params)
+    rng_e, *eval_rngs = jax.random.split(rng_e, n_devices + 1)
+    eval_rngs = jnp.stack(eval_rngs).reshape(n_devices, -1)
+
+    return evaluator, absolute_metric_evaluator, (trained_params, eval_rngs)
+
+
+def run_experiment(_run: Run, _config: Dict, _log: SacredLogger) -> None:
     """Runs experiment."""
+    # Logger setup
+    config = config_copy(_config)
+    log = logger_setup(_run, config, _log)
+
+    scenario_name = config["ENV_SCENARIO"]
+    generator = RandomGenerator(**rware_scenario_configs[scenario_name])
+    # Create envs
+    env = jumanji.make(config["ENV_NAME"], generator=generator)
+    env = RwareMultiAgentWrapper(env)
+    env = AutoResetWrapper(env)
+    env = LogWrapper(env)
+    eval_env = jumanji.make(config["ENV_NAME"])
+    eval_env = RwareMultiAgentWrapper(eval_env)
+
+    # PRNG keys.
+    rng, rng_e, rng_p = jax.random.split(jax.random.PRNGKey(config["SEED"]), num=3)
     # Setup learner.
-    learn, initial_runner_state = learner_setup(env, config)
+    learn, runner_state = learner_setup(env, (rng, rng_p), config)
+
+    # Setup evaluator.
+    evaluator, absolute_metric_evaluator, (trained_params, eval_rngs) = evaluator_setup(
+        eval_env, rng_e, runner_state[0], config
+    )
 
     # Calculate total timesteps.
     n_devices = len(jax.devices())
-    total_timesteps = (
+    config["NUM_UPDATES_PER_EVAL"] = config["NUM_UPDATES"] // config["NUM_EVALUATION"]
+    timesteps_per_training = (
         n_devices
-        * config["NUM_UPDATES"]
+        * config["NUM_UPDATES_PER_EVAL"]
         * config["ROLLOUT_LENGTH"]
         * config["UPDATE_BATCH_SIZE"]
         * config["NUM_ENVS"]
     )
 
-    # Run experiment.
+    # Compile learner and evaluator.
     with TimeIt(tag="COMPILATION"):
-        learn(initial_runner_state)
-
-    with TimeIt(tag="EXECUTION", environment_steps=total_timesteps):
-        learner_output = learn(initial_runner_state)
-
-    return learner_output
+        learn(runner_state)
+    with TimeIt(tag="COMPILATION"):
+        evaluator(trained_params, eval_rngs)
```

Comment:
For @RuanJohn, did we not discuss and say we would remove the explicit compilation step?

Reply:
We removed the extra iteration